### PR TITLE
✨ Add e2e tests scaffolded in default layout

### PR DESCRIPTION
**Pull request by Lauquik**
## Description

adding testing scaffold for the default layout of our project

Test perfoms the follwoing tasks
- Building the Operator Image:
- Loading the Operator Image to Kind:
- Deploying the Controller-Manager:
- Validating Controller-Manager Pod:
## Fixex
#3605

### CODE REVIEW

File: docs/book/src/component-config-tutorial/testdata/project/Makefile [hunk 1]
```diff
@@ -64,6 +64,10 @@ vet: ## Run go vet against code.
 test: manifests generate fmt vet envtest ## Run tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
 
+.PHONY: test-e2e # You will need to have a Kind cluster up in running to run this target
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/Makefile:67]:
```suggestion
# Utilize Kind or modify the e2e tests to load the image locally, enabling compatibility with other vendors.
.PHONY: test-e2e  ## Run the e2e tests against a Kind k8s instance that is spun up.
```

File: docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// Run e2e tests using the Ginkgo runner.
+func TestE2E(t *testing.T) {
+	RegisterFailHandler(Fail)
+	fmt.Fprintf(GinkgoWriter, "Starting Operator suite\n")
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_suite_test.go:14]: We can scaffold the name of the project

File: pkg/plugins/golang/v4/scaffolds/internal/templates/test/e2e/suite.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// Run e2e tests using the Ginkgo runner.
+func TestE2E(t *testing.T) {
+	RegisterFailHandler(Fail)
+	fmt.Fprintf(GinkgoWriter, "Starting Operator suite\n")
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/scaffolds/internal/templates/test/e2e/suite.go:36]:
```suggestion
	fmt.Fprintf(GinkgoWriter, "Starting {{ .ProjectName }} suite\n")
```

We can use the name of the Project
Not all projects built with Kubebuilder are necessary an Operator or is following it.

**Review by camilamacedo86 — COMMENTED** [review]
That's fantastic! 🥇 
Excellent work.

Just a couple of minor nitpicks:
To proceed with the merge, we need to incorporate a GitHub Action that triggers at least one of the tests. It's unnecessary to run tests for all samples—testing just one to confirm its functionality will suffice. For reference, check out this example: [test-sample-go.yml](https://github.com/operator-framework/operator-sdk/blob/master/.github/workflows/test-sample-go.yml).

I think we can consider running the test for `testdata/project-v4`.

**Review by camilamacedo86 — COMMENTED** [review]
Also, please see that we need to have all commits squashed in order to get this one merged.
Thank you again for your amazing contribution 🥇

File: testdata/project-v4-with-grafana/test/utils/utils.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	. "github.com/onsi/ginkgo/v2"
+)
+
```
Comment (camilamacedo86) [inline testdata/project-v4-with-grafana/test/utils/utils.go:11]: It is missing the funcs to install/uninstall cert-manager and Prometheus
See: https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v4-alpha/memcached-operator/test/utils/utils.go#L29-L108

File: testdata/project-v4-with-grafana/test/e2e/e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"sigs.k8s.io/kubebuilder/testdata/project-v4-with-grafana/test/utils"
+)
+
+const namespace = "controller-system"
+
+var _ = Describe("controller", Ordered, func() {
+	Context("Operator", func() {
```
Comment (camilamacedo86) [inline testdata/project-v4-with-grafana/test/e2e/e2e_test.go:17]: It is missing the BeforeAll and AfterAll
BeforeAll should install CertManager and Prometheus
See; https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v4-alpha/memcached-operator/test/e2e/e2e_test.go#L45-L49

Comment (camilamacedo86) [reply testdata/project-v4-with-grafana/test/e2e/e2e_test.go:17]: After all should uninstall all:
https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v4-alpha/memcached-operator/test/e2e/e2e_test.go#L79-L89

File: testdata/project-v4-with-grafana/test/utils/utils.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// Run executes the provided command within this context
+func Run(cmd *exec.Cmd) ([]byte, error) {
+	dir, _ := GetProjectDir()
```
Comment (camilamacedo86) [inline testdata/project-v4-with-grafana/test/utils/utils.go:14]: It is not necessary I think

File: testdata/project-v4-with-grafana/test/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	cmd.Dir = dir
+	fmt.Fprintf(GinkgoWriter, "running dir: %s\n", cmd.Dir)
+
+	// To allow make commands be executed from the project directory which is subdir on SDK repo
+	// TODO:(user) You might not need the following code
+	if err := os.Chdir(cmd.Dir); err != nil {
+		fmt.Fprintf(GinkgoWriter, "chdir dir: %s\n", err)
+	}
```
Comment (camilamacedo86) [inline testdata/project-v4-with-grafana/test/utils/utils.go:22]:
```suggestion
```

i do not think that we need it

File: testdata/project-v4-with-grafana/test/utils/utils.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+// GetProjectDir will return the directory where the project is
+func GetProjectDir() (string, error) {
+	wd, err := os.Getwd()
+	if err != nil {
+		return wd, err
+	}
+	wd = strings.Replace(wd, "/test/e2e", "", -1)
+	return wd, nil
+}
```
Comment (camilamacedo86) [inline testdata/project-v4-with-grafana/test/utils/utils.go:69]:
```suggestion
```

It does not seems required either

File: docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+const namespace = "controller-system"
+
+var _ = Describe("controller", Ordered, func() {
+	Context("Operator", func() {
+		It("should run successfully", func() {
+			var controllerPodName string
+			var err error
+
+			// operatorImage stores the name of the image used in the example
+			var operatorImage = "controller:latest"
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_test.go:23]:
```suggestion
			var projectimage = "projectname:latest"
```

I think we should change here ^

File: docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_test.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+			By("building the manager(Operator) image")
```
Comment (camilamacedo86) [inline docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_test.go:25]: We should not use the word Operator
can you please remove those ones

Comment (Lauquik) [reply docs/book/src/component-config-tutorial/testdata/project/test/e2e/e2e_test.go:25]: alright @camilamacedo86 I'll do all the necessary changes

**Comment by Lauquik** [comment]
@camilamacedo86 to trigger the ci i have wrote one script which will uncomment the cert manager and Prometheus monitor to enable it in the testata/project-v4 and run the e2e-test. iS that okay or need to do anything different?

**Comment by camilamacedo86** [comment]
Hi @Lauquik,

> @camilamacedo86 to trigger the ci i have wrote one script which will uncomment the cert manager and Prometheus monitor to enable it in the testata/project-v4 and run the e2e-test. iS that okay or need to do anything different?

The uncomment cannot be in the scaffold
You can create a Golang code that will use the function that we have to do that see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v4/generate_test.go#L85-L191 (it is exactly what you need to do in the project-v4 prior run the test)

However, if you prefer to do with shell that is fine as well.
The script should be in : https://github.com/kubernetes-sigs/kubebuilder/tree/master/hack
We can cerate a new dir e2e for example

Also, please squash all commits otherwise we cannot get this one merged. 
Thank you a lot 🥇 
You have been doing terrific work !!!

File: Makefile [hunk 11]
```diff
@@ -159,3 +159,7 @@ test-book: ## Run the cronjob tutorial's unit tests to make sure we don't break
 .PHONY: test-license
 test-license:  ## Run the license check
 	./test/check-license.sh
+
+.PHONY: test-e2e-sample-go
+test-e2e-sample-go:
+	./test/e2e/test-e2e-sample.sh
```
Comment (camilamacedo86) [inline Makefile:165]: Nit, we need an blank line at the end of the file

**Comment by Lauquik** [comment]
@camilamacedo86 instead of running new workflow for scaffolded e2e tests. will it be okay to run these test along with test-testdata . i Modifed the test.sh script to run the e2e tests in project-v4

**Comment by Lauquik** [comment]
i think i messed up with this pr, I will create new one

**Pull request closed without merging**